File: AlgoPersonalStudy/src/programmers/PG_12909.java
package programmers;

import java.io.*;
import java.util.*;

public class PG_12909 {
    public static void main(String[] args) {
        String s = "(()(";
        System.out.println(solution(s));
    }
    public static boolean solution(String s){
        boolean answer = true;
        char[] charArr = s.toCharArray();
        if(charArr.length % 2 == 1) return false;

        Stack<Character> stack = new Stack<>();
        for(char c : charArr){
            if(c == '('){
                stack.push(c);
            }
            else{
                if(stack.isEmpty()){
                    answer = false;
                    break;
                }
                stack.pop();
            }
        }
        if(!stack.isEmpty()) answer = false;
        return answer;
    }
}

File: AlgoPersonalStudy/src/_kotlin/programmers/PG_12909.kt
package _kotlin.programmers

import java.util.*

class PG_12909 {
    fun solution(str:String):Boolean{
        var answer = true
        val charArr = str.toCharArray()
        if(charArr.size % 2 == 1) return false

        val stack = Stack<Char>()
        for(c:Char in charArr){
            if(c == '(') stack.push(c)
            else{
                if(stack.isEmpty()){
                    answer = false
                    break
                }
                stack.pop()
            }
        }
        if(!stack.isEmpty()) answer = false
        return answer
    }
}
fun main(){
    val str = "(()("
    val solutionClass = PG_12909()
    println(solutionClass.solution(str))
}
